### PR TITLE
Arduino 101 dts

### PR DESCRIPTION
**Pull request by SavinayDharmappa**

### CODE REVIEW

File: dts/x86/arduino_101.dts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		zephyr,flash = &flash0;
+		zephyr,console = &uart1;
+	};
+};
+
+&uart0 {
+	status = "ok";
+	current-speed = <115200>;
+};
+
```
Comment (jhedberg) [inline dts/x86/arduino_101.dts]: The current arduino_101_defconfig has the following:

 CONFIG_UART_QMSI_0_BAUDRATE=1000000
 CONFIG_UART_QMSI_0_HW_FC=y

Shouldn't arduino_101.dts default to those as well for uart0? Where else will the right values get set? UART0 is the UART between the Quark SE and the nRF51, so it's critical for BLE functionality that the configuration is correct.

File: dts/x86/yaml/arduino101.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+---
+title: ardunio101 Uart
+id: intel,ardunio-uart
+version: 0.1
+
+description: >
+    This binding gives a base representation of the K64 UART
```
Comment (jhedberg) [inline dts/x86/yaml/arduino101.yaml]: K64?

File: dts/x86/arduino_101.dts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/dts-v1/;
+
+#include "intel_curie.dtsi"
+
+/ {
+	model = "Intel Curie Arduino 101 board";
```
Comment (nashif) [inline dts/x86/arduino_101.dts]: just call this Arduino 101, Arduino 101 is not an Intel product.

File: dts/x86/arduino_101.fixup [hunk 4]
```diff
@@ -0,0 +1,5 @@
+
+#define CONFIG_UART_QMSI_1_BAUDRATE	INTEL_ARDUNIO101_UART_B0002400_CURRENT_SPEED
```
Comment (nashif) [inline dts/x86/arduino_101.fixup]: ARDUINO_101_, skip INTEL_

**Review by nashif — APPROVED** [review]

**Review by nashif — APPROVED** [review]
+

**Review by nashif — DISMISSED** [review]
+

File: dts/x86/intel_curie.dtsi [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+		rtc: rtc@B0000400 {
+			compatible = "intel,ardunio101-uart";
+			reg = <0xB0000400 0x400>;
+			clock-frequency = <32768>;
+		};
+
+		uart0: uart@B0002000 {
+			compatible = "intel,ardunio101-uart";
+			reg = <0xB0002000 0x400>;
```
Comment (agross-linaro) [inline dts/x86/intel_curie.dtsi]: A nit: addresses use lower case letters.  So 0xB0002000 should be 0xb0002000.

File: dts/x86/intel_curie.dtsi [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		ranges;
+
+		rtc: rtc@B0000400 {
+			compatible = "intel,ardunio101-uart";
+			reg = <0xB0000400 0x400>;
+			clock-frequency = <32768>;
+		};
+
+		uart0: uart@B0002000 {
+			compatible = "intel,ardunio101-uart";
```
Comment (agross-linaro) [inline dts/x86/intel_curie.dtsi]: mispelling here: intel,arduino101-uart.

File: dts/x86/yaml/arduino101.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+---
+title: ardunio101 Uart
+id: intel,ardunio-uart
```
Comment (agross-linaro) [inline dts/x86/yaml/arduino101.yaml]: Arduino?

File: dts/x86/yaml/arduino101.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+inherits:
+  - !include uart.yaml
+
+properties:
+  - compatible:
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "intel,ardunio101-uart"
```
Comment (agross-linaro) [inline dts/x86/yaml/arduino101.yaml]: Same thing here.  I assume this was supposed to be  arduino.  At least these were all consistent

**Review by agross-oss — CHANGES_REQUESTED** [review]

File: dts/x86/arduino_101.dts (hunk 1 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.dts]: yes it should. To add hw-flowcontrol uart.yaml in has to be changed in dts/common/. I would add uart0 support in arduino.dts in next patch. These set of patch intention was to set the baudrate for uart1 from device tree instead of Kconfig.

File: dts/x86/yaml/arduino101.yaml (hunk 2 above)
Comment (SavinayDharmappa) [inline dts/x86/yaml/arduino101.yaml]: corrected

File: dts/x86/arduino_101.dts (hunk 3 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.dts]: okay.

File: dts/x86/arduino_101.fixup (hunk 4 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.fixup]: okay.

File: dts/x86/intel_curie.dtsi (hunk 6 above)
Comment (SavinayDharmappa) [inline dts/x86/intel_curie.dtsi]: corrected

File: dts/x86/intel_curie.dtsi (hunk 5 above)
Comment (SavinayDharmappa) [inline dts/x86/intel_curie.dtsi]: incorporated the changes

File: dts/x86/yaml/arduino101.yaml (hunk 7 above)
Comment (SavinayDharmappa) [inline dts/x86/yaml/arduino101.yaml]: corrected

File: dts/x86/yaml/arduino101.yaml (hunk 8 above)
Comment (SavinayDharmappa) [inline dts/x86/yaml/arduino101.yaml]: corrected

File: dts/x86/arduino_101.dts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+		zephyr,console = &uart1;
+	};
+};
+
+&uart0 {
+	status = "ok";
+	current-speed = <1000000>;
```
Comment (jhedberg) [inline dts/x86/arduino_101.dts]: This looks better now. What about the HW flow control that was previously enabled through Kconfig? Is that still coming from there?

File: dts/x86/mem.h [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#ifndef __DT_BINDING_ST_MEM_H
+#define __DT_BINDING_ST_MEM_H
+
+#define __SIZE_K(x) (x * 1024)
+
+#define DT_FLASH_SIZE		__SIZE_K(144)
+#define DT_SRAM_SIZE		__SIZE_K(55)
+
```
Comment (tbursztyka) [inline dts/x86/mem.h]: Isn't these valid for arduino only? (I am asking that because the file seems generic to all x86)

File: arch/x86/core/Kconfig [hunk 11]
```diff
@@ -53,12 +53,14 @@ config MAX_IRQ_LINES
 config PHYS_LOAD_ADDR
 	hex "Physical load address"
 	# Default value must be supplied by platform
+	#depends on !HAS_DTS
```
Comment (galak) [inline arch/x86/core/Kconfig]: I think the # before depends is wrong

File: dts/x86/arduino_101.dts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+/dts-v1/;
+
+#include "intel_curie.dtsi"
+
+/ {
+	model = "Arduino 101";
+	compatible = "intel,quark";
```
Comment (galak) [inline dts/x86/arduino_101.dts]: should be: compatible = "arduino,101", "intel,quark";

File: dts/x86/arduino_101.dts (hunk 9 above)
Comment (galak) [inline dts/x86/arduino_101.dts]: I believe HW flow control would come from Kconfig, if we want to do this via DT, we should look to see if the 'uart-has-rtscts' property as defined by Linux Documentation/devicetree/bindings/serial/serial.txt makes sense

File: dts/x86/intel_curie.dtsi [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+#include "mem.h"
+
+/ {
+	cpus {
+		cpu@0 {
+			compatible = "intel,quark";
+		};
+
+		cpu@1 {
+			compatible = "arch";
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: What are you trying to do with compatible = arch?

File: dts/x86/intel_curie.dtsi [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+		};
+	};
+
+	flash0: flash {
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: since at fixed address:

flash0: flash@40010000 {

File: dts/x86/intel_curie.dtsi [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		reg = <0x40010000 DT_FLASH_SIZE>;
+	};
+
+
+	sram0: memory {
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: since at a fixed location:

sram0: memory@a8006400 {

File: dts/x86/yaml/arduino101.yaml [hunk 16]
```diff
@@ -0,0 +1,42 @@
+---
+title: arduino101 Uart
```
Comment (galak) [inline dts/x86/yaml/arduino101.yaml]: Is the uart really arduino101 specific, isn't the UART QMSI?  This should probably be "Intel QMSI Uart" (fixup other uart references accordingly)

File: drivers/serial/Kconfig.qmsi [hunk 17]
```diff
@@ -51,7 +51,7 @@ config UART_QMSI_1_NAME
 	  default name.
 
 config UART_QMSI_1_BAUDRATE
-	depends on UART_QMSI_1
+	depends on !HAS_DTS
```
Comment (galak) [inline drivers/serial/Kconfig.qmsi]: You should get the 'name' and 'baudrate' for all QMSI UARTs from the DT.

File: dts/x86/arduino_101.fixup [hunk 18]
```diff
@@ -1,7 +1,5 @@
-#define CONFIG_UART_QMSI_0_BAUDRATE	INTEL_ARDUNIO101_UART_B0002000_CURRENT_SPEED
```
Comment (galak) [inline dts/x86/arduino_101.fixup]: Why are these removed in this commit?

**Review by galak — CHANGES_REQUESTED** [review]
A number of cleanups / changes needed.

File: dts/x86/intel_curie.dtsi (hunk 13 above)
Comment (SavinayDharmappa) [inline dts/x86/intel_curie.dtsi]: It should be arc. I will correct it.

File: dts/x86/mem.h (hunk 10 above)
Comment (SavinayDharmappa) [inline dts/x86/mem.h]: It is specific to Quark SE C1000.

File: dts/x86/arduino_101.dts (hunk 12 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.dts]: okay.

File: dts/x86/intel_curie.dtsi (hunk 14 above)
Comment (SavinayDharmappa) [inline dts/x86/intel_curie.dtsi]: okay

File: dts/x86/intel_curie.dtsi (hunk 15 above)
Comment (SavinayDharmappa) [inline dts/x86/intel_curie.dtsi]: okay

File: dts/x86/arduino_101.fixup (hunk 18 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.fixup]: without sanity check i pushed for first time. It caused warning in build so removed and pushed it agin

File: drivers/serial/Kconfig.qmsi (hunk 17 above)
Comment (SavinayDharmappa) [inline drivers/serial/Kconfig.qmsi]: okay

File: dts/x86/arduino_101.fixup (hunk 18 above)
Comment (galak) [inline dts/x86/arduino_101.fixup]: ok, little weird, lets get the other things in order than we can figure this out.

File: dts/x86/arduino_101.dts (hunk 9 above)
Comment (SavinayDharmappa) [inline dts/x86/arduino_101.dts]: yes HW flowcontrol is comming from Kconfig. DTS support of  HW flowcontrol requires changes in  dts/common/yaml/uart.yaml so i will push those changes as new patch.

File: dts/x86/yaml/arduino101.yaml [hunk 19]
```diff
@@ -0,0 +1,42 @@
+---
```
Comment (galak) [inline dts/x86/yaml/arduino101.yaml]: Add this in the patch to updates the serial driver.

File: dts/x86/yaml/arduino101.yaml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+inherits:
+  - !include uart.yaml
+
+properties:
+  - compatible:
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "arduino_101-uart"
```
Comment (galak) [inline dts/x86/yaml/arduino101.yaml]: Why is this arduino_101-uart, and not intel,qmsi-uart?

File: dts/x86/intel_curie.dtsi [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+		uart0: uart@b0002000 {
+			compatible = "arduino_101-uart";
+			reg = <0xb0002000 0x400>;
+			label = "UART_0";
+
+			status = "disabled";
+		};
+
+		uart1: uart@b0002400 {
+			compatible = "arduino_101-uart";
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: compatiable = "intel,qmsi-uart"?

File: dts/x86/intel_curie.dtsi [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+		ranges;
+
+		rtc: rtc@b0000400 {
+			compatible = "arduino_101-rtc";
+			reg = <0xb0000400 0x400>;
+			clock-frequency = <32768>;
+		};
+
+		uart0: uart@b0002000 {
+			compatible = "arduino_101-uart";
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: compatiable = "intel,qmsi-uart"?

File: drivers/serial/Kconfig.qmsi [hunk 23]
```diff
@@ -43,15 +44,15 @@ config UART_QMSI_1
 	default n
 
 config UART_QMSI_1_NAME
-	depends on UART_QMSI_1
+	depends on !HAS_DTS
```
Comment (galak) [inline drivers/serial/Kconfig.qmsi]: Leave the depends on as is, lets do the same if !HAS_DTS

**Review by galak — CHANGES_REQUESTED** [review]
Some changes for UART, lets split out the flow control patch into a separate PR.

**Review by galak — CHANGES_REQUESTED** [review]
In general things are looking good, however I would restructure the patches so we have:
1. changes to serial driver
2. changes to x86 in general
3. changes that add arduino 101 support
(anything to boards/x86/arduino_101 or has arduino_101 in the name).

**Review by galak — APPROVED** [review]
